open_spiel/games/battleship_test.cc

```diff
@@ -0,0 +1,261 @@
+  // vertically.
+
+  const auto game =
+      LoadGame("battleship", {{"board_width", GameParameter(3)},
+                              {"board_height", GameParameter(2)},
+                              {"ship_sizes", GameParameter("[2;3]")},
+                              {"ship_values", GameParameter("[1;1]")}});
+  auto state = game->NewInitialState();
+  SPIEL_CHECK_EQ(state->CurrentPlayer(), Player{0});
+  {
+    const auto actions = state->LegalActions();
+    SPIEL_CHECK_EQ(actions, std::vector<Action>({0, 1, 3, 4}));
+    SPIEL_CHECK_EQ(state->ActionToString(actions[0]), "h_0_0");
+    SPIEL_CHECK_EQ(state->ActionToString(actions[1]), "h_0_1");
+    SPIEL_CHECK_EQ(state->ActionToString(actions[2]), "h_1_0");
+    SPIEL_CHECK_EQ(state->ActionToString(actions[3]), "h_1_1");
+  }
+
+  // We now place the first ship to the right on the first row, which
+  // corresponds to the serialized index 1 as checked above.
+  state->ApplyAction(Action{1});
+
+  // We repeat the check for Player 1.
+  SPIEL_CHECK_EQ(state->CurrentPlayer(), Player{1});
+  {
+    const auto actions = state->LegalActions();
+    SPIEL_CHECK_EQ(actions, std::vector<Action>({0, 1, 3, 4}));
+    SPIEL_CHECK_EQ(state->ActionToString(actions[0]), "h_0_0");
+    SPIEL_CHECK_EQ(state->ActionToString(actions[1]), "h_0_1");
+    SPIEL_CHECK_EQ(state->ActionToString(actions[2]), "h_1_0");
+    SPIEL_CHECK_EQ(state->ActionToString(actions[3]), "h_1_1");
+  }
+
+  // We place Player 1's ship to the left on the second row.
+  state->ApplyAction(Action{3});
+
+  // Now, we need to check that the only remaining action for Player 0 is to
+  // place the second ship on the second row.
+  SPIEL_CHECK_EQ(state->CurrentPlayer(), Player{0});
+  {
+    const auto actions = state->LegalActions();
+    SPIEL_CHECK_EQ(actions, std::vector<Action>({3}));
+    SPIEL_CHECK_EQ(state->ActionToString(actions[0]), "h_1_0");
+  }
+  state->ApplyAction(Action{3});
+
+  //... While for Player 1 the only remaining action is to place the second ship
+  // on the first row.
+  SPIEL_CHECK_EQ(state->CurrentPlayer(), Player{1});
+  {
+    const auto actions = state->LegalActions();
+    SPIEL_CHECK_EQ(actions, std::vector<Action>({0}));
+    SPIEL_CHECK_EQ(state->ActionToString(actions[0]), "h_0_0");
+  }
+  state->ApplyAction(Action{0});
+
+  SPIEL_CHECK_FALSE(state->IsTerminal());
+}
+
+void TestNashEquilibriumInSmallBoard() {
+  // We replicate the same setting as page 7 of [1].
+  //
+  // There, each player has a 1x3 board with a single ship of size 1 and value
+  // 1.0. Each player has two shots available. The loss multiplier is 2.0, so
+  // this is a *general-sum* game.
+  //
+  // The only Nash equilibrium of the game is for all players to place their
+  // ship at random, and then shoot at random.
+  //
+  //
+  // [1]:
+  // https://papers.nips.cc/paper/9122-correlation-in-extensive-form-games-saddle-point-formulation-and-benchmarks.pdf#page=7
```

Comment:
Ah, this is awesome. Thanks a lot for doing this!

Reply:
Thanks for the enthusiasm! As soon as OpenSpiel implements algorithms for exact optimal correlated equilibria, we can add more to the test :-)